File: TickTackToe.py
import KI
import numpy as np
import timeit
"""
Player  1
Player -1
"""
#checks wether one player has won returns player or 0 if none

def check_winner(map):
    for i in range(3):
     if (map[i][0] == map[i][1] and map[i][1] == map[i][2]):
      return map[i][0]
    
    for i in range(3):
     if (map[0][i] == map[1][i] and map[1][i] == map[2][i]):
      return map[0][i]
     
    if (map[0][0] == map[1][1] and map[1][1] == map[2][2]):
     return map[0][0]
     
    if (map[0][2] == map[1][1] and map[1][1] == map[2][0]):
     return map[1][1]
    
    else:
     return 0

def print_map(map):
    print(map[0])
    print(map[1])
    print(map[2])

#converts map to array for ai

def conv_to_array(map, player):
    state = []
    for i in range(3):
     for j in range(3):
      if (map[i][j] == player):
       state.append(1)
      elif (map[i][j] == 0):
       state.append(0)
      else:
       state.append(-1)
    return state

def validate_play(play,map):
    if (map[int(play /3)][play %3] == 0):
     return 1
    else:
     return 0  

def find_valid_play(play,map):
    while 1:
     if (validate_play(np.argmax(play),map) == 0):
      play[np.argmax(play)] = 0
     else:
      return play

def playing(ki_1, ki_2,stable):
    map = [[0] * 3 for i in range(3)]
    i = 0
    while 1:
     play = KI.brain(stable[ki_1], conv_to_array(map,1))
     i = i + 1
     play = find_valid_play(play,map)
     map[int(np.argmax(play)/3)][np.argmax(play)%3] = 1
     if (not check_winner(map) == 0):
      return check_winner(map)
     if (i == 9):
      return 0

     play = KI.brain(stable[ki_2], conv_to_array(map,-1))
     i = i + 1
     play = find_valid_play(play,map)
     map[int(np.argmax(play)/3)][np.argmax(play)%3] = -1
     if (not check_winner(map) == 0):
      return check_winner(map)

def playing_human(ki_1,stable):
    map = [[0] * 3 for i in range(3)]
    i = 0
    while 1:
     play = KI.brain(stable[ki_1], conv_to_array(map,1))
     i = i + 1
     play = find_valid_play(play,map)
     map[int(np.argmax(play)/3)][np.argmax(play)%3] = 1
     if (not check_winner(map) == 0):
      return check_winner(map)
     if (i == 9):
      return 0
     
     print_map(map)
     play = int(input("Input: "))
     i = i + 1
     map[int(play/3)][play%3] = -1
     if (not check_winner(map) == 0):
      return check_winner(map)
 

Agents = 100
Generations = 1000
Trainer = 10  #x best networks play against the rest

stop = 1 # used for first timing because i am lazy
start = 1

stable = KI.Stable_init(Agents,9,9,7,9)
fit = [0]*Agents
for x in range(Generations):
 print("Playing Round ", x ,"of ", Generations, " Estimated Time: ", round((stop-start)*(Generations-x)/60, 1 ), "min")
 start = timeit.default_timer()
 for i in range(Trainer):
  for j in range(Agents):
   a = playing(i,j,stable)
   if (a == 1):
    fit[i] = fit[i] + 1
   if (a == -1):
    fit[j] = fit[j] + 1
 KI.Stable_Knecht(fit,stable,0.01,0.2,0.1)
 stop = timeit.default_timer()
KI.save_stable(stable)
playing_human(0,stable)

File: TickTackToe.c
#include "KI.h"
#include "stdlib.h"
#include "stdio.h"

int* init_game(){
    int* gamefield = malloc(9*sizeof(int));
    for (int i = 0; i < 9; i++){
	gamefield[i] = 0;
    }
    return gamefield;
}

int check_winner(int* gamefield){
    for (int i = 0; i < 3;i++){
	if (gamefield[i*3] == gamefield[i*3+1] && gamefield[i*3] == gamefield[i*3+2]){
	    return gamefield[i*3];
	}
	if (gamefield[i] == gamefield[i+3] && gamefield[i]== gamefield[i+6]){
	    return gamefield[i];
	}
    }
    if (gamefield[0] == gamefield[4] && gamefield[0] == gamefield[8]){
        return gamefield[0];
    }
    if (gamefield[2] == gamefield[4] && gamefield[2] == gamefield[6]){
        return gamefield[2];
    } 
    return 0;
}


void print_map(int* gamefield){
    for (int i = 0; i < 3;i++ ){
	printf("%i %i %i\n",gamefield[i*3],gamefield[i*3+1],gamefield[i*3+2]);
    }
}

int validate_play(int play,int* gamefield){
    if (gamefield[play] == 0){
	return 1;
    }
    else{
	return 0;
    }
}

int* sort_play(float* play){
    int* sort = malloc(sizeof(int)*9);
    for (int i = 0; i < 9;i++){
	sort[i] = i; 
    }
    for (int i = 0; i < 9;i++){
	for (int j = 0;j < 8;j++){
	    if (play[j] < play[j+1]){
		float savef = play[j+1];
		play[j+1] = play[j];
		play[j] = savef;
		
		int save;
		save = sort[j+1];
		sort[j+1] = sort[j];
		sort[j] = save;
	    }
	}
    }
    return sort;
}

void find_play(int* gamefield, float* play, int player){
    int* sort = sort_play(play);
    for (int i = 0; i < 9; i++){
	if (validate_play(sort[i], gamefield)){
	    gamefield[sort[i]] = player;
	    free(sort);
	    i = 9;
	}
    }
} 

float* random_play(){
    float* play = malloc(sizeof(float)*9);
    for (int i = 0; i < 9; i++){
	play[i] = random();
    }
    return play;
}

float playing(genom_parameter gen_para,float* genom){
    int* gamefield = init_game();
    float* play;
    int outcome;
    int i = 0;
    while (1){
	play = random_play();
	find_play(gamefield, play,1);
	free(play);
	outcome = check_winner(gamefield);
	if (outcome != 0){
	    free(gamefield);
	    return 0.0;
	}
	i++;
	if(i == 9){
	    free(gamefield);
	    return 0.0;
	}
	play = brain(gen_para, genom, (float*) gamefield);
	find_play(gamefield, play, 2);
	free(play);
	
	outcome = check_winner(gamefield);
	if (outcome != 0){
	    free(gamefield);
	    return 1.0;
	}
	i++;
    }
}

int main(){
    int size = 1000;
    genom_parameter gen_para;
    init_genom_parameter(&gen_para, 9, 9, 9, 2, size);
    float** stable = stable_init(gen_para);
    float fitness[size];
    float average;
    for (int l = 0; l < 10000; l++){
	average = 0;
	for (int i = 0; i<size;i++){
    	    fitness[i] = 0;
    	    for (int j = 0;j<100;j++){
    	        fitness[i] = fitness[i] + playing(gen_para, stable[i]);
    	    }
	    average = fitness[i] + average;
    	}
    	stable_knecht(gen_para, stable, fitness, 0.005, 0.01, 0.1, size/2);
	printf("Generation: %i Average Fitness: %f\n", l, average/size);
    }
    free(stable);
}
